### PR TITLE
opencascade-occt_7-4 init version freeze at 7.4.0p1

### PR DESCRIPTION
**Pull request by cafkafk**
###### Description of changes
This PR seeks to create a version freeze at 7.4 to help an effort into packaging topologic with nixpkgs.

Since topologic is dependent on 7.4, there needs to exist a package freeze.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/opencascade-occt_7-4/default.nix [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  libGL, libGLU, libXext, libXmu, libXi, darwin }:
+
+stdenv.mkDerivation rec {
+  pname = "opencascade-occt_7.4";
+  version = "7.4.0p1";
+  commit = "V${builtins.replaceStrings ["."] ["_"] version}";
+
+  src = fetchurl {
+    name = "occt-${commit}.tar.gz";
+    url = "https://git.dev.opencascade.org/gitweb/?p=occt.git;a=snapshot;h=${commit};sf=tgz";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/opencascade-occt_7-4/default.nix:11]:
```suggestion
  commit = "";

  src = fetchurl {
    url = "https://git.dev.opencascade.org/gitweb/?p=occt.git;a=snapshot;h=V${builtins.replaceStrings ["."] ["_"] version};sf=tgz";
```

File: pkgs/development/libraries/opencascade-occt_7-4/default.nix [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    description = "Open CASCADE Technology, libraries for 3D modeling and numerical simulation";
+    homepage = "https://www.opencascade.org/";
+    license = licenses.lgpl21;  # essentially...
+    # The special exception defined in the file OCCT_LGPL_EXCEPTION.txt
+    # are basically about making the license a little less share-alike.
+    maintainers = with maintainers; [ cafkafk ];
+    platforms = platforms.all;
+  };
+
+}
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/opencascade-occt_7-4/default.nix:29]:
```suggestion
  };
}
```

**Comment by SuperSandro2000** [comment]
> This PR seeks to create a version freeze at 7.4 to help an effort into packaging topologic with nixpkgs.

Since the package is otherwise identically topologic should just use a local overwrite.

File: pkgs/development/libraries/opencascade-occt_7-4/default.nix (hunk 1 above)
Comment (cafkafk) [reply pkgs/development/libraries/opencascade-occt_7-4/default.nix:11]: My idea was to keep it as close as possible to the current opencascade-occt package. Perhaps it should be changed there first?

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -20748,6 +20748,8 @@ with pkgs;
   };
   opencascade-occt = callPackage ../development/libraries/opencascade-occt { };
 
+  opencascade-occt_7-4 = callPackage ../development/libraries/opencascade-occt_7-4 { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:20751]:
```suggestion
  opencascade-occt = callPackage ../development/libraries/opencascade-occt { };

  opencascade-occt_7-4 = callPackage ../development/libraries/opencascade-occt/7.4nix { };
```

**Comment by SuperSandro2000** [comment]
> I agree that it would be fine for topologic. My impression is that a lot of in-house solutions still relied on this version, and I thought packaging it separately would mean those users could use nix to install it. That would include using it for those things that are not packaged yet (including topologic which might not be easy to package).
> 
> As I understand it, a lot of the construction industry suffers immensely from a lack of reproducible builds and lack of technical know how, and I think nix can solve a lot of problems for them. So I was sort of hoping to use this as a little pilot example.
> 
> I dunno if that justifies making a separate version, and I'm perfectly fine with it not doing so in nixpkgs if it causes too many issues.

So it is just a hypothetical use case? nixpkgs tries to keep as little versions of software as possible to reduce maintenance burden of old and dated versions of software.

@amiloradovsky @gebner do you think we need this extra version?

**Comment by gebner** [comment]
I think we should only add the pinned 7.4 version if it is required for a package in nixpkgs. So when adding `topologic` to nixpkgs, it would be okay to add the pinned 7.4 version of opencascade.

If we add it, it should be in a `opencascade-occt/7.4.nix` file instead of a separate directory.

**Comment by cafkafk** [comment]
> So it is just a hypothetical use case? nixpkgs tries to keep as little versions of software as possible to reduce maintenance burden of old and dated versions of software.

I know someone who is currently doing gsoc that would have an issue solved if this was merged.

I'll happily commit to maintaining it, and if it causes any tree wide issues I'm fine with dropping it. Ideally it would be temporary like was done with e.g. xmonad_0_17_0



> I think we should only add the pinned 7.4 version if it is required for a package in nixpkgs. So when adding `topologic` to nixpkgs, it would be okay to add the pinned 7.4 version of opencascade.
> 
> If we add it, it should be in a `opencascade-occt/7.4.nix` file instead of a separate directory.

I mean it could be a part of a topologic pr then, but I'll probably not have that done anytime soon.

If it's not able to be merged before I'll just have people clone the branch I'm pushing from and install
it with that until then I suppose.

**Pull request closed without merging**